compiler-plugin/src/io/ktor/compiler/utils/ScopedExpressionEvaluator.kt

```diff
@@ -0,0 +1,294 @@
+
+            // Get the symbol for this property access
+            val calleeReference = transformed.calleeReference as? FirResolvedNamedReference ?: return transformed
+            val symbol = calleeReference.resolvedSymbol as? FirCallableSymbol<*> ?: return transformed
+
+            // If the symbol is in our scope, replace with its value
+            if (variables.containsVariable(symbol)) {
+                val value = variables.getVariableValue(symbol) ?: return transformed
+
+                // Create a copy of the value that preserves type information from the original expression
+                return when (value) {
+                    is FirLiteralExpression -> {
+                        buildLiteralExpression(
+                            source = transformed.source,
+                            kind = value.kind,
+                            value = value.value,
+                            annotations = transformed.annotations.toMutableList(),
+                            setType = false
+                        ).apply {
+                            replaceConeTypeOrNull(transformed.resolvedType)
+                        }
+                    }
+                    else -> {
+                        // For non-literal expressions, we might need to clone them
+                        // while preserving type information from the original
+                        value.transformChildren(this, data)
+                    }
+                } as FirStatement
+            }
+
+            return transformed
+        }
+    }
+
+    /**
+     * Resolves type variables within a ConeTypeProjection based on the current type scope
+     */
+    fun resolveTypeProjection(projection: ConeTypeProjection): ConeKotlinType? {
+        // If it's already a type, just make sure any type parameters in it are resolved
+        if (projection is ConeKotlinType) {
+            return resolveType(projection)
+        }
+
+        // Handle different kinds of projections
+        return when (projection) {
+            is ConeKotlinTypeProjection -> {
+                val resolvedType = resolveType(projection.type)
+
+                // Apply projection kind if needed
+                when (projection.kind) {
+                    ProjectionKind.INVARIANT -> resolvedType
+                    ProjectionKind.IN -> resolvedType?.let { ConeKotlinTypeProjectionIn(it).type }
+                    ProjectionKind.OUT -> resolvedType?.let { ConeKotlinTypeProjectionOut(it).type }
+                    ProjectionKind.STAR -> null // Star projection can't be resolved to a concrete type
+                }
+            }
+            is ConeStarProjection -> null // Star projection can't be resolved to a concrete type
+            else -> null // Unknown projection type
+        }
+    }
+
+    /**
+     * Resolves type parameters within a type using the current type scope
+     */
+    fun resolveType(type: ConeKotlinType): ConeKotlinType? {
```

Comment:
Shouldn't there be a utility that does that? Or maybe it is resolved in another phase?